### PR TITLE
feat: ✨ sd-switch

### PR DESCRIPTION
**Pull request by DanielHargesheimer**

### CODE REVIEW

File: packages/components/src/components/switch/switch.ts [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+ */
+@customElement('sd-switch')
+export default class SdSwitch extends SolidElement implements SolidFormControl {
+  private readonly formControlController = new FormControlController(this, {
+    value: (control: SdSwitch) => (control.checked ? control.value || 'on' : undefined),
+    defaultValue: (control: SdSwitch) => control.defaultChecked,
+    setValue: (control: SdSwitch, checked: boolean) => (control.checked = checked)
+  });
+
+  @query('input[type="checkbox"]') input: HTMLInputElement;
```
Comment (mariohamann) [inline packages/components/src/components/switch/switch.ts:41]:
```suggestion
  @query('input') input: HTMLInputElement;
```

should be enough

File: packages/components/src/components/switch/switch.ts [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+          @blur=${this.handleBlur}
+          @focus=${this.handleFocus}
+        />
+
+        <span
+          id="control"
+          part="control ${this.checked ? ' control--checked' : 'control--unchecked'}"
+          class=${cx(
+            `relative flex flex-initial items-center justify-center border rounded-full h-4 w-8  
+            peer-focus-visible:outline peer-focus-visible:outline-2 peer-focus-visible:outline-offset-2
```
Comment (mariohamann) [inline packages/components/src/components/switch/switch.ts:183]: Wasn't aware of `peer`. Interesting. @yoezlem @abudd1094

File: packages/components/src/components/switch/switch.ts [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+    this.input.setCustomValidity(message);
+    this.formControlController.updateValidity();
+  }
+
+  render() {
+    return html`
+      <label
+        part="base"
+        class=${cx(
+          'sd-switch group flex items-center text-base leading-normal text-black cursor-pointer',
```
Comment (mariohamann) [inline packages/components/src/components/switch/switch.ts:157]:
```suggestion
          'group flex items-center text-base leading-normal text-black cursor-pointer',
```

**Comment by mariohamann** [comment]
Please add color transitions.

**Comment by DanielHargesheimer** [comment]
> Please add color transitions.

Hmmm, I tried it, but it looked so odd. One time the color transition looked to be to late and then to early. I think, to animate just the movement of the circle is the best solution. What do you think @coraliefeil?

**Comment by mariohamann** [comment]
Sorry, I meant especially for hover and focus states. :)

Maybe it has to come fast for the checked state? Could you try that with a very short transition like duration-50?

**Comment by coraliefeil** [comment]
Left a comment on chromatic ... reassign when updated. Thanks a lot!

File: packages/components/src/components/switch/switch.stories.ts [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    return generateTemplate({
+      axis: {
+        y: { type: 'attribute', name: 'required' }
+      },
+      args
+    });
+  }
+};
+
+export const Checked = {
```
Comment (yoezlem) [inline packages/components/src/components/switch/switch.stories.ts:71]: I am missing the description for that

File: packages/components/src/components/switch/switch.stories.ts [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+  },
+  play: async ({ canvasElement }: { canvasElement: HTMLUnknownElement }) => {
+    const el = canvasElement.querySelector('sd-button');
+    await waitUntil(() => el?.shadowRoot?.querySelector('button'));
+    await userEvent.type(el!.shadowRoot!.querySelector('button')!, '{return}', { pointerEventsCheck: 0 });
+  }
+};
+
+/**
+ * Use the `base`, `control--unchecked`, `control--checked`, `checked` and `label` part selectors to customize the checkbox.
```
Comment (yoezlem) [inline packages/components/src/components/switch/switch.stories.ts:113]: I don't know if I would go with the naming checkbox 🤔 I would Propose to take the name from the component -> switch

File: packages/components/src/components/switch/switch.stories.ts [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+ */
+
+export const Default = {
+  render: (args: any) => {
+    return generateTemplate({ args });
+  }
+};
+
+/**
+ * Use the disabled attribute to disable an input checkbox. Clicks will be suppressed until the disabled state is removed
```
Comment (yoezlem) [inline packages/components/src/components/switch/switch.stories.ts:35]: I don't know if I would go with the naming checkbox 🤔 I would Propose to take the name from the component -> switch

File: packages/components/src/components/switch/switch.stories.ts [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+            };
+          })
+        }
+      },
+      args
+    });
+  }
+};
+
+const checkboxTemplate = (part: string) => {
```
Comment (yoezlem) [inline packages/components/src/components/switch/switch.stories.ts:163]: It might be stupid, but no matter how easy something looks, we shouldn't forget to describe it 🙂

File: packages/components/src/components/switch/switch.test.ts [hunk 8]
```diff
@@ -1,17 +1,21 @@
-import { aTimeout, expect, fixture, html, oneEvent, waitUntil } from '@open-wc/testing';
+import { expect, fixture, html, oneEvent, waitUntil } from '@open-wc/testing';
 import { sendKeys } from '@web/test-runner-commands';
 import sinon from 'sinon';
 import type SdSwitch from './switch';
 
 describe('<sd-switch>', () => {
   it('should pass accessibility tests', async () => {
-    const el = await fixture<SdSwitch>(html` <sd-switch>Switch</sd-switch> `);
+    const el = await fixture<SdSwitch>(html` <sd-switch>Checkbox</sd-switch> `);
```
Comment (yoezlem) [inline packages/components/src/components/switch/switch.test.ts:8]: Why did you rename the "label"?

File: packages/components/src/components/switch/switch.ts [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+import { ifDefined } from 'lit/directives/if-defined.js';
+import { live } from 'lit/directives/live.js';
+import { property, query } from 'lit/decorators.js';
+import { watch } from '../../internal/watch';
+import cx from 'classix';
+import SolidElement from '../../internal/solid-element';
+import type { SolidFormControl } from '../../internal/solid-element';
+
+/**
+ * @summary Checkboxes allow the user to toggle an option on or off.
```
Comment (yoezlem) [inline packages/components/src/components/switch/switch.ts:14]:
```suggestion
 * @summary Switches allow the user to toggle an option on or off.
```

File: packages/components/src/components/switch/switch.ts [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+export default class SdSwitch extends SolidElement implements SolidFormControl {
+  private readonly formControlController = new FormControlController(this, {
+    value: (control: SdSwitch) => (control.checked ? control.value || 'on' : undefined),
+    defaultValue: (control: SdSwitch) => control.defaultChecked,
+    setValue: (control: SdSwitch, checked: boolean) => (control.checked = checked)
+  });
+
+  @query('input[type="checkbox"]') input: HTMLInputElement;
+
+  @property() title = ''; // make reactive to pass through
```
Comment (yoezlem) [inline packages/components/src/components/switch/switch.ts:43]: Please add a description to the title

**Comment by yoezlem** [comment]
the interactive states (hover) aren't working..
image

File: packages/components/src/components/switch/switch.stories.ts (hunk 5 above)
Comment (DanielHargesheimer) [reply packages/components/src/components/switch/switch.stories.ts:113]: But switch or switched is not a state like checked. Then you have to say switched-on, switched-off. I can do that.

Comment (yoezlem) [reply packages/components/src/components/switch/switch.stories.ts:113]: I am talking about the last word of your description:
image

I would change that text to: "**part selectors to customize the switch.**"

Comment (yoezlem) [reply packages/components/src/components/switch/switch.stories.ts:113]: The rest with control--unchecked, control--checked is good! :)

**Comment by coraliefeil** [comment]
Approved, thank you so much.

**Review by abudd1094 — APPROVED** [review]
I like the invalid story, I handled it slightly differently in input.  Everything is working on your branch and requirements are met.  Nice one.

**Comment by yoezlem** [comment]
Looks good to me!

**Review by yoezlem — APPROVED** [review]
